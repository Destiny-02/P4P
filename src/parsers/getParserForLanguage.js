// @ts-check
const javaParser = require("./languages/java");
const javascriptParser = require("./languages/javascript");
const pythonParser = require("./languages/python");

/** @param {string} fileName */
const getParserForLanguage = (fileName) =>
  fileName.endsWith(".java")
    ? javaParser
    : fileName.endsWith(".js")
    ? javascriptParser
    : fileName.endsWith(".py")
    ? pythonParser
    : undefined;

module.exports = { getParserForLanguage };
