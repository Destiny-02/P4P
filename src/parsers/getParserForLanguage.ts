// @ts-check

import { JavaParser } from "./languages/java";
import { JsTsParser } from "./languages/javascript";
import { PythonParser } from "./languages/python";

export const getParserForLanguage = (fileName: string) =>
  fileName.endsWith(".java")
    ? JavaParser
    : /.[ajt]sx?$/.test(fileName)
    ? JsTsParser
    : fileName.endsWith(".py")
    ? PythonParser
    : undefined;
