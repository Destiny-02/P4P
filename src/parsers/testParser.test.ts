import { readdirSync } from "node:fs";
import { join } from "node:path";
import { getParserForLanguage } from "./getParserForLanguage";

const testCaseFolderPath = join(__dirname, "tests");
const testCaseFileNames = readdirSync(testCaseFolderPath);

// this file automatically creates test cases for every file in the "tests"
// directory. We shouldn't need to edit this file.

describe("test case files", () => {
  for (const fileName of testCaseFileNames) {
    const Parser = getParserForLanguage(fileName);
    if (Parser) {
      // generate a test case for each file

      const parser = new Parser();

      it(`parses ${fileName} (${parser.language})`, async () => {
        const pathToFile = join(testCaseFolderPath, fileName);
        expect(await parser.parse(pathToFile)).toMatchSnapshot();
      });
    } else {
      it.todo(`parses ${fileName} (no parser available)`);
    }
  }
});
