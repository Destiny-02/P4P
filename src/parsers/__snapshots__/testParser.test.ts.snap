// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test case files parses Sample.java (java) 1`] = `
{
  "comments": [
    "// import from another package",
    "// import from our own package",
    "/**
   * Sample is a class
   */",
    "// this is a comment",
    "// these are all globals that should not be considered",
  ],
  "identifiers": [
    {
      "isLibraryName": false,
      "name": "Sample",
      "sourceLocation": [
        367,
        372,
      ],
      "type": "typeIdentifier",
      "typeDefinition": {
        "typeName": "class",
      },
    },
    {
      "isLibraryName": false,
      "name": "get_annual_income_deduction",
      "sourceLocation": [
        424,
        450,
      ],
      "type": "methodDeclarator",
      "typeDefinition": {
        "argumentTypes": [
          "Map",
          "int",
        ],
        "returnType": "int",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "myFirstArg",
      "sourceLocation": [
        473,
        482,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "public",
        ],
        "typeName": "Map",
      },
    },
    {
      "isLibraryName": false,
      "name": "secondArg",
      "sourceLocation": [
        489,
        497,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "public",
        ],
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "a",
      "sourceLocation": [
        510,
        510,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "public",
        ],
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "b",
      "sourceLocation": [
        525,
        525,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "public",
        ],
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "employees",
      "sourceLocation": [
        558,
        566,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "public",
        ],
        "typeName": "List",
      },
    },
    {
      "isLibraryName": false,
      "name": "paye_tax_bracket",
      "sourceLocation": [
        668,
        683,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "private",
          "static",
          "final",
        ],
        "typeName": "double",
      },
    },
    {
      "isLibraryName": false,
      "name": "threadId",
      "sourceLocation": [
        733,
        740,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "public",
          "static",
          "volatile",
          "transient",
        ],
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "get_payslip_for_paye",
      "sourceLocation": [
        783,
        802,
      ],
      "type": "methodDeclarator",
      "typeDefinition": {
        "argumentTypes": [
          "int",
          "String",
          "int",
        ],
        "returnType": "String",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "arg1",
      "sourceLocation": [
        808,
        811,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "protected",
          "synchronized",
        ],
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "arg2",
      "sourceLocation": [
        821,
        824,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "protected",
          "synchronized",
        ],
        "typeName": "String",
      },
    },
    {
      "isLibraryName": false,
      "name": "spreadArguments",
      "sourceLocation": [
        834,
        848,
      ],
      "type": "variableArityParameter",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": [
          "protected",
          "synchronized",
        ],
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "MyEnum",
      "sourceLocation": [
        979,
        984,
      ],
      "type": "typeIdentifier",
      "typeDefinition": {
        "typeName": "enum",
      },
    },
    {
      "isLibraryName": false,
      "name": "ENUM_MEMBER_1",
      "sourceLocation": [
        992,
        1004,
      ],
      "type": "enumConstant",
      "typeDefinition": {
        "typeName": "intrinsic",
      },
    },
    {
      "isLibraryName": false,
      "name": "ENUM_MEMBER_2",
      "sourceLocation": [
        1011,
        1023,
      ],
      "type": "enumConstant",
      "typeDefinition": {
        "typeName": "intrinsic",
      },
    },
    {
      "isLibraryName": false,
      "name": "MyIFace",
      "sourceLocation": [
        1043,
        1049,
      ],
      "type": "typeIdentifier",
      "typeDefinition": {
        "typeName": "interface",
      },
    },
    {
      "isLibraryName": false,
      "name": "myInterfaceProperty",
      "sourceLocation": [
        1061,
        1079,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "int",
      },
    },
    {
      "isLibraryName": false,
      "name": "i",
      "sourceLocation": [
        1151,
        1151,
      ],
      "type": "variableDeclaratorId",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "int",
      },
    },
  ],
  "imports": {
    "named": {
      "ArrayList": {
        "isLibraryFile": true,
        "originalName": "ArrayList",
        "source": "java.util",
      },
      "List": {
        "isLibraryFile": true,
        "originalName": "List",
        "source": "java.util",
      },
      "SmartCardReader": {
        "isLibraryFile": false,
        "originalName": "SmartCardReader",
        "source": "nz.school.app.cordovaSmartCardPlugin",
      },
      "Whatever": {
        "isLibraryFile": false,
        "originalName": "Whatever",
        "source": "parsers.languages",
      },
    },
    "wildcard": [
      {
        "isLibraryFile": true,
        "source": "java.beans",
      },
      {
        "isLibraryFile": false,
        "source": "nz.school.app.cordovaGeoLocationPlugin",
      },
      {
        "isLibraryFile": false,
        "source": "parsers.languages",
      },
    ],
  },
}
`;

exports[`test case files parses complex.py (python) 1`] = `
{
  "comments": [
    "
This is a block comment in the middle of nowhere

it has a fake # inline comment within the block

",
    "block comment above func - this is not valid",
    "block comment within func",
    "# comment on the same line as the block end",
  ],
  "identifiers": [
    {
      "isLibraryName": false,
      "name": "identifier1",
      "sourceLocation": [
        211,
        222,
      ],
      "type": "FunctionDeclaration.ownId",
    },
    {
      "isLibraryName": false,
      "name": "identifier2",
      "sourceLocation": [
        264,
        275,
      ],
      "type": "VariableDeclarator.ownId",
    },
    {
      "isLibraryName": false,
      "name": "identifier3",
      "sourceLocation": [
        292,
        303,
      ],
      "type": "FunctionDeclaration.ownId",
    },
    {
      "isLibraryName": false,
      "name": "tripleQuotedString",
      "sourceLocation": [
        318,
        336,
      ],
      "type": "VariableDeclarator.ownId",
    },
  ],
  "imports": {
    "named": {},
    "wildcard": [],
  },
}
`;

exports[`test case files parses sample.as (javascript) 1`] = `
{
  "comments": [
    " eslint-disable ",
    " @ts-nocheck -- this is a sample file, not part of the sourcecode",
    " an inline comment",
    " these are all globals that should not be considered",
  ],
  "identifiers": [
    {
      "isLibraryName": false,
      "name": "a",
      "sourceLocation": [
        103,
        104,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "any",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "b",
      "sourceLocation": [
        106,
        107,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "any",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "add",
      "sourceLocation": [
        99,
        102,
      ],
      "type": "FunctionDeclaration.ownId",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "any",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        160,
        161,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        160,
        161,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        171,
        172,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        171,
        172,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "z",
      "sourceLocation": [
        182,
        183,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "z",
      "sourceLocation": [
        182,
        183,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "SomeClass",
      "sourceLocation": [
        228,
        237,
      ],
      "type": "ClassDeclaration.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
  ],
  "imports": {
    "named": {},
    "wildcard": [],
  },
}
`;

exports[`test case files parses sample.js (javascript) 1`] = `
{
  "comments": [
    " eslint-disable ",
    " @ts-nocheck -- this is a sample file, not part of the sourcecode",
    "*
 * comments yay
 * @param {number} a
 * @param {number} b
 ",
    " an inline comment",
    " these are all globals that should not be considered",
  ],
  "identifiers": [
    {
      "isLibraryName": false,
      "name": "a",
      "sourceLocation": [
        560,
        561,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "any",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "b",
      "sourceLocation": [
        563,
        564,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "any",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "add",
      "sourceLocation": [
        556,
        559,
      ],
      "type": "FunctionDeclaration.ownId",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "any",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        617,
        618,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        617,
        618,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        628,
        629,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        628,
        629,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "z",
      "sourceLocation": [
        639,
        640,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "z",
      "sourceLocation": [
        639,
        640,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "SomeClass",
      "sourceLocation": [
        685,
        694,
      ],
      "type": "ClassDeclaration.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "i",
      "sourceLocation": [
        823,
        824,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "j",
      "sourceLocation": [
        830,
        831,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "i",
      "sourceLocation": [
        823,
        824,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "j",
      "sourceLocation": [
        830,
        831,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x1",
      "sourceLocation": [
        866,
        868,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x1",
      "sourceLocation": [
        866,
        868,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x2",
      "sourceLocation": [
        888,
        890,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x2",
      "sourceLocation": [
        888,
        890,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "item",
      "sourceLocation": [
        917,
        921,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "index",
      "sourceLocation": [
        923,
        928,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "undefined",
      "sourceLocation": [
        957,
        966,
      ],
      "type": "ArrowFunctionExpression.undefined",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        948,
        949,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "i",
      "sourceLocation": [
        951,
        952,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "accumulator",
      "sourceLocation": [
        979,
        990,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "item",
      "sourceLocation": [
        992,
        996,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "index",
      "sourceLocation": [
        998,
        1003,
      ],
      "type": "ArrowFunctionExpression.argument",
      "typeDefinition": {
        "classification": "index",
        "modifiers": undefined,
        "typeName": "any",
      },
    },
  ],
  "imports": {
    "named": {
      "defaultImport": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile.js",
      },
      "defaultImport2": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
      "json": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./foo.json",
      },
      "named1": {
        "isLibraryFile": false,
        "originalName": "named1",
        "source": "./someFile",
      },
      "named2": {
        "isLibraryFile": false,
        "originalName": "named2",
        "source": "./someFile",
      },
      "named3": {
        "isLibraryFile": false,
        "originalName": "named3",
        "source": "./someFile.ts",
      },
      "named5": {
        "isLibraryFile": false,
        "originalName": "named5",
        "source": "./someFile",
      },
      "renamed4": {
        "isLibraryFile": false,
        "originalName": "named4",
        "source": "./someFile.ts",
      },
      "renamed5": {
        "isLibraryFile": false,
        "originalName": "named5",
        "source": "./someFile",
      },
      "renamedDefault": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
      "renamedDefault2": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
    },
    "wildcard": [],
  },
}
`;

exports[`test case files parses sample.md (md) 1`] = `
{
  "comments": [
    "# Sample Documentation

This is a sample markdown file, which serves as documentation for SAAAP (Sample Acronym and Abbreviation Parser).

Some abbreivations are less straightforward, like COM (communication port) or starboard (STBD).

Abbreviations might also exist in lists:

- VTS: Vessel Traffic Services
- AIS ( Automatic Identification System )
- UTC: Universal Coordinated Time
- European Train Control System (ETCS)
- COSPAS-SARSAT (Cosmicheskaya Sistema Poiska Avariynyh Sudov - Search And Rescue Satellite-Aided Tracking)
",
  ],
  "identifiers": [],
  "imports": {
    "named": {},
    "wildcard": [],
  },
}
`;

exports[`test case files parses sample.py (python) 1`] = `
{
  "comments": [
    "
Hiii
",
    "
    does nothing
    ",
    "# inline commenttt",
    "# these are all globals that should not be considered",
  ],
  "identifiers": [
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        28,
        29,
      ],
      "type": "FunctionDeclaration.argument",
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        31,
        32,
      ],
      "type": "FunctionDeclaration.argument",
    },
    {
      "isLibraryName": false,
      "name": "subtract",
      "sourceLocation": [
        19,
        27,
      ],
      "type": "FunctionDeclaration.ownId",
    },
    {
      "isLibraryName": false,
      "name": "a",
      "sourceLocation": [
        74,
        75,
      ],
      "type": "VariableDeclarator.ownId",
    },
    {
      "isLibraryName": false,
      "name": "OOP",
      "sourceLocation": [
        88,
        91,
      ],
      "type": "FunctionDeclaration.ownId",
    },
    {
      "isLibraryName": false,
      "name": "doSomething",
      "sourceLocation": [
        135,
        146,
      ],
      "type": "MemberExpression.property",
    },
    {
      "isLibraryName": false,
      "name": "str1",
      "sourceLocation": [
        169,
        173,
      ],
      "type": "VariableDeclarator.ownId",
    },
    {
      "isLibraryName": false,
      "name": "str2",
      "sourceLocation": [
        228,
        232,
      ],
      "type": "VariableDeclarator.ownId",
    },
  ],
  "imports": {
    "named": {},
    "wildcard": [],
  },
}
`;

exports[`test case files parses sample.ts (javascript) 1`] = `
{
  "comments": [
    " eslint-disable ",
    " @ts-nocheck -- this is a sample file, not part of the sourcecode",
    " extra: TS only",
    "* a comment  ",
    " another comment",
    " these are all globals that should not be considered",
    " no type argument for the first parameter",
  ],
  "identifiers": [
    {
      "isLibraryName": false,
      "name": "a",
      "sourceLocation": [
        715,
        724,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "number",
          "number",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "b",
      "sourceLocation": [
        726,
        735,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "number",
          "number",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "add",
      "sourceLocation": [
        711,
        714,
      ],
      "type": "FunctionDeclaration.ownId",
      "typeDefinition": {
        "argumentTypes": [
          "number",
          "number",
        ],
        "modifiers": undefined,
        "returnType": "any",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        769,
        770,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "x",
      "sourceLocation": [
        769,
        770,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        780,
        781,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "y",
      "sourceLocation": [
        780,
        781,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "z",
      "sourceLocation": [
        791,
        858,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "string | (number & {}) | (string & string) | Map<number, string>",
      },
    },
    {
      "isLibraryName": false,
      "name": "z",
      "sourceLocation": [
        791,
        858,
      ],
      "type": "VariableDeclarator.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "string | (number & {}) | (string & string) | Map<number, string>",
      },
    },
    {
      "isLibraryName": false,
      "name": "SomeClass",
      "sourceLocation": [
        903,
        912,
      ],
      "type": "ClassDeclaration.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "MyIFace",
      "sourceLocation": [
        1060,
        1067,
      ],
      "type": "TSInterfaceDeclaration.ownId",
      "typeDefinition": undefined,
    },
    {
      "isLibraryName": false,
      "name": "MyType",
      "sourceLocation": [
        1098,
        1104,
      ],
      "type": "TSTypeAliasDeclaration.ownId",
      "typeDefinition": undefined,
    },
    {
      "isLibraryName": false,
      "name": "ENUM_MEMBER_1",
      "sourceLocation": [
        1149,
        1162,
      ],
      "type": "TSEnumMember.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "ENUM_MEMBER_2",
      "sourceLocation": [
        1166,
        1179,
      ],
      "type": "TSEnumMember.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "MyEnum",
      "sourceLocation": [
        1138,
        1144,
      ],
      "type": "TSEnumDeclaration.ownId",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "any",
      },
    },
    {
      "isLibraryName": false,
      "name": "a",
      "sourceLocation": [
        1375,
        1376,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "never",
          "number[]",
        ],
        "modifiers": undefined,
        "returnType": "ReturnType",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "b",
      "sourceLocation": [
        1378,
        1386,
      ],
      "type": "FunctionDeclaration.argument",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "never",
          "number[]",
        ],
        "modifiers": undefined,
        "returnType": "ReturnType",
        "typeName": "function",
      },
    },
    {
      "isLibraryName": false,
      "name": "spread",
      "sourceLocation": [
        1391,
        1397,
      ],
      "type": "RestElement.argument",
      "typeDefinition": {
        "classification": undefined,
        "modifiers": undefined,
        "typeName": "number[]",
      },
    },
    {
      "isLibraryName": false,
      "name": "myOtherFunction",
      "sourceLocation": [
        1359,
        1374,
      ],
      "type": "FunctionDeclaration.ownId",
      "typeDefinition": {
        "argumentTypes": [
          "any",
          "never",
          "number[]",
        ],
        "modifiers": undefined,
        "returnType": "ReturnType",
        "typeName": "function",
      },
    },
  ],
  "imports": {
    "named": {
      "Type1": {
        "isLibraryFile": false,
        "originalName": "Type1",
        "source": "./someFile",
      },
      "Type2": {
        "isLibraryFile": false,
        "originalName": "Type2",
        "source": "./someFile",
      },
      "Type3": {
        "isLibraryFile": false,
        "originalName": "Type3",
        "source": "./someFile",
      },
      "defaultImport": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile.js",
      },
      "defaultImport2": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
      "defaultWithType": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
      "defaultWithType2": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
      "json": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./foo.json",
      },
      "named1": {
        "isLibraryFile": false,
        "originalName": "named1",
        "source": "./someFile",
      },
      "named2": {
        "isLibraryFile": false,
        "originalName": "named2",
        "source": "./someFile",
      },
      "named3": {
        "isLibraryFile": false,
        "originalName": "named3",
        "source": "./someFile.ts",
      },
      "named5": {
        "isLibraryFile": false,
        "originalName": "named5",
        "source": "./someFile",
      },
      "named6": {
        "isLibraryFile": false,
        "originalName": "named6",
        "source": "./someFile",
      },
      "renamed4": {
        "isLibraryFile": false,
        "originalName": "named4",
        "source": "./someFile.ts",
      },
      "renamed5": {
        "isLibraryFile": false,
        "originalName": "named5",
        "source": "./someFile",
      },
      "renamedDefault": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
      "renamedDefault2": {
        "isLibraryFile": false,
        "originalName": "default",
        "source": "./someFile",
      },
    },
    "wildcard": [],
  },
}
`;
